docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+	NginxContextHTTPServerLocation NginxContext = "http.server.location"
+
+	// NginxContextStream is the stream context of the NGINX configuration.
+	NginxContextStream NginxContext = "stream"
+
+	// NginxContextStreamServer is the server context of the NGINX configuration.
+	NginxContextStreamServer NginxContext = "stream.server"
+)
+
+// Snippet represents an NGINX configuration snippet.
+type Snippet struct {
+	// Context is the NGINX context to insert the snippet into.
+	Context NginxContext `json:"context"`
+
+	// Value is the NGINX configuration snippet.
+	Value string `json:"value"`
+}
+```
+
+#### Supported NGINX Contexts
+
+SnippetsPolicy supports the following NGINX configuration contexts:
+
+- `main`
+- http module: `http`, `server`, `location`
+- stream module: `stream`, `server`
+
+`upstream` context for both http and stream modules is not included for the following reasons:
+
+- http module - there are not enough features in https://nginx.org/en/docs/http/ngx_http_upstream_module.html
+  - There are a few `keepalive`-related directives. But they also require `proxy_set_header Connection "";`, which
+    is possible to configure with location snippets, but it will conflict with the generated Connection header (see
+    https://github.com/nginxinc/nginx-gateway-fabric/blob/5968bc348213a8470f6aaaa1a9bd51f2e90523ac/internal/mode/static/nginx/config/servers.go#L39-L40
+    and https://github.com/nginxinc/nginx-gateway-fabric/blob/5968bc348213a8470f6aaaa1a9bd51f2e90523ac/internal/mode/static/nginx/config/maps_template.go#L21-L26)
+    so those keepalive directives won't work.
+  - Session persistence ([`sticky`](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#sticky)). Those are
+    valid use cases. But at the same time, they only apply to NGINX Plus. Additionally,
+    Gateway API started introducing session persistence.
+    See https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.BackendLBPolicy
+- stream module - there are not any features in https://nginx.org/en/docs/stream/ngx_stream_upstream_module.html
+
+Note: because NGF already inserts the `random` load-balancing method, an `upstream` snippet will not be able to
+configure
+a different method.
+
+If we choose to introduce upstream snippets, the SnippetsPolicy will need to support targeting a Service, because NGF
+generates one upstream per Service.
+
+#### Examples
+
+Below are a few examples of using the SnippetsPolicy to bring unsupported NGINX configuration into NGF.
+
+##### Rate-limiting
+
+We use NGINX [limit req module](https://nginx.org/en/docs/http/ngx_http_limit_req_module.html) to configure rate
+limiting.
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsPolicy
+metadata:
+  name: rate-limit
+spec:
+  targetRefs:
+  - group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: cafe-route
+  snippets:
+  - context: http
+    value: limit_req_zone $binary_remote_addr zone=myzone:10m rate=1r/s;
```

Comment:
I'm assuming this config could also be attached to a Gateway resource, but the location config below could not. Could we allow multiple policies where this is defined on the Gateway, and then the location section is defined on a route?

Or do your conflict rules below prevent this case? If so, how do we enforce it? Or maybe since snippets are somewhat freeform, we can't really enforce it and it's in the user's hands.

Reply:
this should work fine with the Gateway.

NGF insert the`http` snippet into http context (only once) and the location snippet per each location of HTTPRoutes that belong to the Gateway.